pkg/reconciler/v1alpha1/taskrun/validate_resources.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either extress or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package taskrun
+
+import (
+	"fmt"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/knative/build-pipeline/pkg/reconciler/v1alpha1/taskrun/list"
+	"github.com/knative/build-pipeline/pkg/reconciler/v1alpha1/taskrun/resources"
+)
+
+func validateInputResources(inputs *v1alpha1.Inputs, providedResources map[string]*v1alpha1.PipelineResource) error {
+	if inputs != nil {
+		return validateResources(inputs.Resources, providedResources)
+	}
+	return validateResources([]v1alpha1.TaskResource{}, providedResources)
+}
+
+func validateOutputResources(outputs *v1alpha1.Outputs, providedResources map[string]*v1alpha1.PipelineResource) error {
+	if outputs != nil {
+		return validateResources(outputs.Resources, providedResources)
+	}
+	return validateResources([]v1alpha1.TaskResource{}, providedResources)
+}
+
+func validateResources(neededResources []v1alpha1.TaskResource, providedResources map[string]*v1alpha1.PipelineResource) error {
+	needed := make([]string, 0, len(neededResources))
+	for _, resource := range neededResources {
+		needed = append(needed, resource.Name)
+	}
+	provided := make([]string, 0, len(providedResources))
+	for resource := range providedResources {
+		provided = append(provided, resource)
+	}
+	missing := list.DiffLeft(needed, provided)
+	if len(missing) > 0 {
+		return fmt.Errorf("missing resources: %s", missing)
```

Comment:
Is there a way to surface this error with word `input`/`output`? I think it would be helpful to understand for the user what to fix in their yaml

eg: 
```
missing input resources a,b
```

Reply:
good call!

i think that we will get this currently, though it will look a bit ugly - `validateResources` will return the error `missing resources: foo`, which will be caught by `ValidateResolvedTaskResources`, which will prepend `invalid output resources: ` to the error, so in the end I think we'll end up with:

`invalid output resources: missing resources: foo`

Which is a bit icky but I think does have all the info we need.

(on a related note I think we should consider exploring wrapping our errors, e.g. using https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully !)